ee/agent/storage/bbolt/backup.go

```diff
@@ -0,0 +1,173 @@
+package agentbbolt
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/kolide/launcher/ee/agent/types"
+	"github.com/kolide/launcher/pkg/launcher"
+	"go.etcd.io/bbolt"
+)
+
+const (
+	backupInitialDelay = 10 * time.Minute
+	backupInterval     = 1 * time.Hour
+)
+
+// databaseBackupSaver periodically takes backups of launcher.db.
+type databaseBackupSaver struct {
+	knapsack    types.Knapsack
+	slogger     *slog.Logger
+	interrupt   chan struct{}
+	interrupted bool
+}
+
+func NewDatabaseBackupSaver(k types.Knapsack) *databaseBackupSaver {
+	return &databaseBackupSaver{
+		knapsack:  k,
+		slogger:   k.Slogger().With("component", "database_backup_saver"),
+		interrupt: make(chan struct{}, 1),
+	}
+}
+
+func (d *databaseBackupSaver) Execute() error {
+	// Wait a little bit after startup before taking first backup, to allow for enrollment
+	select {
+	case <-d.interrupt:
+		d.slogger.Log(context.TODO(), slog.LevelDebug,
+			"received external interrupt during initial delay, stopping",
+		)
+		return nil
+	case <-time.After(backupInitialDelay):
+		break
+	}
+
+	// Take periodic backups
+	ticker := time.NewTicker(backupInterval)
+	defer ticker.Stop()
+	for {
+		if err := d.backupDb(); err != nil {
+			d.slogger.Log(context.TODO(), slog.LevelWarn,
+				"could not perform periodic database backup",
+				"err", err,
+			)
+		}
+
+		select {
+		case <-ticker.C:
+			continue
+		case <-d.interrupt:
+			d.slogger.Log(context.TODO(), slog.LevelDebug,
+				"interrupt received, exiting execute loop",
+			)
+			return nil
+		}
+	}
+}
+
+func (d *databaseBackupSaver) Interrupt(_ error) {
+	// Only perform shutdown tasks on first call to interrupt -- no need to repeat on potential extra calls.
+	if d.interrupted {
+		return
+	}
+	d.interrupted = true
+
+	d.interrupt <- struct{}{}
+}
+
+func (d *databaseBackupSaver) backupDb() error {
+	// Take backup -- it's fine to just overwrite previous backups
+	backupLocation := BackupLauncherDbLocation(d.knapsack.RootDirectory())
+	if err := d.knapsack.BboltDB().View(func(tx *bbolt.Tx) error {
+		return tx.CopyFile(backupLocation, 0600)
+	}); err != nil {
+		return fmt.Errorf("backing up database: %w", err)
+	}
+
+	// Confirm file exists and is nonempty
+	if exists, err := launcher.NonEmptyFileExists(backupLocation); !exists {
+		return fmt.Errorf("backup succeeded, but nonempty file does not exist at %s", backupLocation)
+	} else if err != nil {
+		return fmt.Errorf("backup succeeded, but error checking if file was created at %s: %w", backupLocation, err)
+	}
+
+	// Log success
+	d.slogger.Log(context.TODO(), slog.LevelDebug,
```

Comment:
I'd log this from the caller on `err=nil` or whatever.

Reply:
(ended up moving this back due to the extra rotate/rename steps)